File: dart/phone_number.dart
class PhoneNumber {
  String clean(String number) {
    if (RegExp(r'[a-zA-Z]').hasMatch(number)) {
      throw FormatException('letters not permitted');
    }

    if (RegExp(r'[^-+\s.()0-9]').hasMatch(number)) {
      throw FormatException('punctuations not permitted');
    }

    number = number.replaceAll(RegExp(r'[^0-9]'), '');

    if (number.length > 11) {
      throw FormatException('more than 11 digits');
    }

    if (number.length == 11) {
      if (number[0] != '1') {
        throw FormatException('11 digits must start with 1');
      }
      number = number.substring(1);
    }

    if (number.length != 10) {
      throw FormatException('incorrect number of digits');
    }

    if (number[0] == '0') {
      throw FormatException('area code cannot start with zero');
    }

    if (number[0] == '1') {
      throw FormatException('area code cannot start with one');
    }

    if (number[3] == '0') {
      throw FormatException('exchange code cannot start with zero');
    }

    if (number[3] == '1') {
      throw FormatException('exchange code cannot start with one');
    }

    return number;
  }
}

File: ocaml/phone_number.ml
open Base

let check s =
  if Char.equal s.[0] '0' then Error "area code cannot start with zero"
  else if Char.equal s.[0] '1' then Error "area code cannot start with one"
  else if Char.equal s.[3] '0' then Error "exchange code cannot start with zero"
  else if Char.equal s.[3] '1' then Error "exchange code cannot start with one"
  else Ok s

let number s =
  let is_valid = String.contains "+-.() 0123456789" in
  if String.exists s ~f:Char.is_alpha then Error "letters not permitted"
  else if String.exists s ~f:(Fn.compose not is_valid) then
    Error "punctuations not permitted"
  else
    let t = String.filter s ~f:Char.is_digit in
    match String.length t with
    | len when len > 11 -> Error "more than 11 digits"
    | 10 -> check t
    | 11 when Char.equal t.[0] '1' -> check @@ String.drop_prefix t 1
    | 11 -> Error "11 digits must start with 1"
    | _ -> Error "incorrect number of digits"
